Objects/typeobject.c

```diff
@@ -3078,19 +3079,28 @@ type_new_descriptors(const type_new_ctx *ctx, PyTypeObject *type)
         }
     }
 
-    if (ctx->add_dict && ctx->base->tp_itemsize) {
-        type->tp_dictoffset = -(long)sizeof(PyObject *);
-        slotoffset += sizeof(PyObject *);
+    if (ctx->add_dict) {
+        if (ctx->base->tp_itemsize) {
+            type->tp_dictoffset = -(long)sizeof(PyObject *);
+            slotoffset += sizeof(PyObject *);
+        }
+        else {
+            assert((type->tp_flags & Py_TPFLAGS_MANAGED_DICT) == 0);
+            type->tp_flags |= Py_TPFLAGS_MANAGED_DICT;
+        }
     }
 
     if (ctx->add_weak) {
         assert(!ctx->base->tp_itemsize);
-        type->tp_weaklistoffset = slotoffset;
-        slotoffset += sizeof(PyObject *);
+        if (type->tp_flags & Py_TPFLAGS_MANAGED_DICT) {
```

Comment:
Could `Py_TPFLAGS_MANAGED_DICT` be set even if wasn't set on line 3089?  In other words, could the flag have been set before `type_new_descriptors()` was called?

If so, that implies that `Py_TPFLAGS_MANAGED_DICT` also implies a managed weaklist, which I'm not sure is correct.

Otherwise the assert on line 3088 should probably move up to before line 3082.

Reply:
Further, if it's always either both-are-managed or neither then it would be helpful if the flag were more clear (e.g. `Py_TPFLAGS_MANAGED_DICT_AND_WEAKLIST`).

If either can be managed separately then there should probably be a separate `Py_TPFLAGS_MANAGED_WEAKLIST` flag.

That might be necessary anyway since `Py_TPFLAGS_MANAGED_DICT` is already part of 3.11.

Reply:
We want both to be managed, or neither, but our hand might be forced in the case of multiple inheritance.
One base might have a managed dict, and the other an explicit weakreflist.
I expect this case to be rare.

Currently `Py_TPFLAGS_MANAGED_DICT` is internal, it is set by the VM, but shouldn't be set by third-party code.

When we make it public, as we should do for 3.12, then it would make sense to add `Py_TPFLAGS_MANAGED_WEAKLIST`.

I'd like some sort of decision on https://github.com/python/cpython/issues/95589 first, though.

Reply:
Thinking about this a bit more, we can ensure both are explicit or both managed.
If one is explicit, the we make both explicit.

Moving from an explicitly declare offset to a managed one is unsafe, so we can't do that.
But we can do it the other way around.

Reply:
Here it only matters in the context of `type_new_descriptors()`.  If we can guarantee the both-or-neither aspect in this "limited" context then that is sufficient for this PR (assuming we leave a comment to that effect).

Reply:
Regardless, my original question stands: could `Py_TPFLAGS_MANAGED_DICT` already be set when `type_new_descriptors()` is called?  If so, the branch on 3095 is now problematic, no?